### ISSUE TITLE
unsorted index error in GRUAggregation and LSTMAgrregation 

### ISSUE THREAD

**Issue by farzad1132** [pyg-team/pytorch_geometric#7642]
### 
In a nutshell, the problem is that `GRUAgrregation` and `LSTMAggregation` require sorted `index`, and whenever our input `edge_index` to these models is arranged in a way that the resulting `index` is not sorted, we will get the following error:

`Can not perform aggregation since the 'index' tensor is not sorted`  Compete traceback is at the end of this issue

Using the code below, I will elaborate on the bug and a possible solution.

## Code to reproduce the bug
```python
import os.path as osp
import torch as th
from torch_geometric.datasets import IMDB
from torch_geometric.nn import SAGEConv, to_hetero
from torch_geometric.nn.aggr import GRUAggregation

# importing a sample heterogeneous dataset
path = osp.join(osp.dirname(osp.realpath(__file__)), "data/")
dataset = IMDB(path)
data = dataset[0]


# Code sample from https://pytorch-geometric.readthedocs.io/en/latest/tutorial/heterogeneous.html#automatically-converting-gnn-models
# with one small change (I have changed aggregation scheme of conv1 to GRU)
class GNN(th.nn.Module):
    def __init__(self, hidden_channels, in_channels, out_channels):
        super().__init__()
        self.conv1 = SAGEConv(
            (-1, -1), hidden_channels, aggr=GRUAggregation(in_channels, hidden_channels)
        )
        self.conv2 = SAGEConv((-1, -1), out_channels)

    def forward(self, x, edge_index):
        x = self.conv1(x, edge_index).relu()
        x = self.conv2(x, edge_index)
        return x


model = GNN(
    hidden_channels=64, out_channels=2, in_channels=dataset.num_features["movie"]
)
model = to_hetero(model, data.metadata(), aggr="sum")


result = model(data.x_dict, data.edge_index_dict) # you will see the error here
```

## Details of the bug
This bug comes from the fact that these aggregators use `to_dense_batch` utility, which requires input `index` to be sorted, as we can see below, which is a part of `Aggregation` implementation.

```python
def to_dense_batch(
        self,
        x: Tensor,
        index: Optional[Tensor] = None,
        ptr: Optional[Tensor] = None,
        dim_size: Optional[int] = None,
        dim: int = -2,
        fill_value: float = 0.,
        max_num_elements: Optional[int] = None,
    ) -> Tuple[Tensor, Tensor]:

        # TODO Currently, `to_dense_batch` can only operate on `index`:
        self.assert_index_present(index)
        self.assert_sorted_index(index)
        self.assert_two_dimensional_input(x, dim)

        return to_dense_batch(
            x,
            index,
            batch_size=dim_size,
            fill_value=fill_value,
            max_num_nodes=max_num_elements,
        )
```

## One possible solution
Since the order of `edge_index` doesn't matter at all, we can use `sort_edge_index` utility to sort the `edge_index` **column-wise** because `index` directly comes from columns of `edge_index`.

**Note that the order in `edge_index` doesn't change the input graph at all**

So with this little change, we can solve the problem:

```python

from torch_geometric.utils import sort_edge_index

data.edge_index_dict = {key: sort_edge_index(value, sort_by_row=False) 
                        for key, value in data.edge_index_dict.items()}
result = model(data.x_dict, data.edge_index_dict)
```

In my opinion, the following changes must be made:

1. An integrated solution must be developed.
2. Other aggregators must be checked to ascertain whether they have a similar bug or not.
3. If none of the above two done, at least, we should update the documentation to inform users.

## Traceback
```
Can not perform aggregation since the 'index' tensor is not sorted
  File "/home/farzad/venvs/celnet-3.8/lib/python3.8/site-packages/torch_geometric/nn/aggr/base.py", line 132, in assert_sorted_index
    raise ValueError("Can not perform aggregation since the 'index' "
  File "/home/farzad/venvs/celnet-3.8/lib/python3.8/site-packages/torch_geometric/nn/aggr/base.py", line 170, in to_dense_batch
    self.assert_sorted_index(index)
  File "/home/farzad/venvs/celnet-3.8/lib/python3.8/site-packages/torch_geometric/nn/aggr/gru.py", line 35, in forward
    x, _ = self.to_dense_batch(x, index, ptr, dim_size, dim)
  File "/home/farzad/venvs/celnet-3.8/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/farzad/venvs/celnet-3.8/lib/python3.8/site-packages/torch_geometric/nn/aggr/base.py", line 109, in __call__
    return super().__call__(x, index, ptr, dim_size, dim, **kwargs)
  File "/home/farzad/venvs/celnet-3.8/lib/python3.8/site-packages/torch_geometric/nn/conv/message_passing.py", line 608, in aggregate
    return self.aggr_module(inputs, index, ptr=ptr, dim_size=dim_size,
  File "/home/farzad/venvs/celnet-3.8/lib/python3.8/site-packages/torch_geometric/nn/conv/message_passing.py", line 484, in propagate
    out = self.aggregate(out, **aggr_kwargs)
  File "/home/farzad/venvs/celnet-3.8/lib/python3.8/site-packages/torch_geometric/nn/conv/sage_conv.py", line 131, in forward
    out = self.propagate(edge_index, x=x, size=size)
  File "/home/farzad/venvs/celnet-3.8/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "<eval_with_key>.1", line 14, in forward
    conv1__director = self.conv1.movie__to__director((x__movie, x__director), edge_index__movie__to__director)
  File "/home/farzad/venvs/celnet-3.8/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/farzad/venvs/celnet-3.8/lib/python3.8/site-packages/torch/fx/graph_module.py", line 277, in __call__
    raise e
  File "/home/farzad/venvs/celnet-3.8/lib/python3.8/site-packages/torch/fx/graph_module.py", line 277, in __call__
    raise e
  File "/home/farzad/venvs/celnet-3.8/lib/python3.8/site-packages/torch/fx/graph_module.py", line 658, in call_wrapped
    return self._wrapped_call(self, *args, **kwargs)
  File "/home/farzad/projects/GNN/practice/test2.py", line 47, in <module>
    result = model(data.x_dict, data.edge_index_dict)
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main (Current frame)
    return _run_code(code, main_globals, None,
ValueError: Can not perform aggregation since the 'index' tensor is not sorted
```

### Environment

* PyG version:  2.3.0
* PyTorch version: 1.13.1+cu117
* OS: Ubuntu 22.04 LTS
* Python version: 3.8
* CUDA/cuDNN version: 11.7
* How you installed PyTorch and PyG (`conda`, `pip`, source): pip
* Any other relevant information (*e.g.*, version of `torch-scatter`):

**Comment by rusty1s** [comment]
Yes, your understand is correct. Can you clarify what you mean by
> An integrated solution must be developed.
I am happy to make this process better.

> Other aggregators must be checked to ascertain whether they have a similar bug or not.
What do you mean by bug? Whenever an `Aggregation` requires sorted `index` tensor, this error will be thrown right now.
> If none of the above two done, at least, we should update the documentation to inform users.
Sounds good.

How about the following: We add `Data.sort()` and `HeteroData.sort()` functionality to PyG, and additionally inform the user better on how to resolve this issue in the error message and in the documentation?

**Comment by rusty1s** [comment]
Will be resolved with #7649
